source/extensions/filters/http/credential_injector/credential_injector_filter.cc

```diff
@@ -0,0 +1,94 @@
+#include "source/extensions/filters/http/credential_injector/credential_injector_filter.h"
+
+#include "source/common/http/headers.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace CredentialInjector {
+
+FilterConfig::FilterConfig(CredentialInjectorSharedPtr credential_injector, bool overwrite,
+                           bool allow_request_without_credential, const std::string& stats_prefix,
+                           Stats::Scope& scope)
+    : injector_(credential_injector), overwrite_(overwrite),
+      allow_request_without_credential_(allow_request_without_credential),
+      stats_(generateStats(stats_prefix + "credential_injector.", scope)) {}
+
+CredentialInjectorFilter::CredentialInjectorFilter(FilterConfigSharedPtr config)
+    : config_(std::move(config)) {}
+
+Http::FilterHeadersStatus CredentialInjectorFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                                  bool) {
+  ENVOY_LOG(debug, "Called Filter : {}", __func__);
+
+  request_headers_ = &headers;
+
+  in_flight_credential_request_ = config_->requestCredential(*this);
+
+  // pause while we await the next step from the credential source, for example, an OAuth server
+  return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+}
+
+void CredentialInjectorFilter::onSuccess() {
+  decoder_callbacks_->dispatcher().post([this]() {
```

Comment:
Can you explain why we need to `post`?

Reply:
Are you saying `onSuccess` is executed on the main worker? But then you are accessing `decode_callbacks_` concurrently from two threads in the failure case? Don't you need to post at the top level for any data access to decoder callbacks?

I assume you plan to add some sort of shared cache between workers to handle remote fetching? Can you explain how this would work with the cache.

Reply:
I think it might be better to assume callbacks are called locally and push the coordination into the credential provider. For example, I took at a look at `gcp_authn` and it handles the cache with TLS but all callbacks are thread-local. CC @tyxia

Reply:
@kyessenov credential providers may need to fetch the credentials remotely, to avoid blocking the work thread, this needs to be done in a separate thread. Therefore, when providers succeed or fail to get the credential, they use the `onSuccess` or `onFailure` callback to notify the credential injector filter.

The `onFailure` callback should also post the handling to the work thread.

Reply:
I think you can remove this post now since the credential supplies manages the threads itself.